scipy/optimize/lbfgsb.py

```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
     relationship between the two is ``ftol = factr * numpy.finfo(float).eps``.
     I.e., `factr` multiplies the default machine floating-point precision to
     arrive at `ftol`.
-
+    If any of the lower bounds are equal to the upper bounds then the
+    corresponding entry in ``OptimizeResult.jac`` will be np.nan.
     """
     _check_unknown_options(unknown_options)
     m = maxcor
     pgtol = gtol
     factr = ftol / np.finfo(float).eps
 
-    x0 = asarray(x0).ravel()
-    n, = x0.shape
+    full_x0 = asarray(x0).ravel()
 
     if bounds is None:
-        bounds = [(None, None)] * n
-    if len(bounds) != n:
+        bounds = [(-np.inf, np.inf)] * full_x0.size
+    if len(bounds) != full_x0.size:
         raise ValueError('length of x0 != length of bounds')
 
-    # unbounded variables must use None, not +-inf, for optimizer to work properly
-    bounds = [(None if l == -np.inf else l, None if u == np.inf else u) for l, u in bounds]
     # LBFGSB is sent 'old-style' bounds, 'new-style' bounds are required by
     # approx_derivative and ScalarFunction
-    new_bounds = old_bound_to_new(bounds)
+    full_new_bounds = old_bound_to_new(bounds)
+
+    # initial vector must lie within the bounds. Otherwise ScalarFunction and
+    # approx_derivative will cause problems
+    full_x0 = np.clip(full_x0, full_new_bounds[0], full_new_bounds[1])
```

Comment:
Was this an existing bug that you are fixing here?

Reply:
It's [existing behaviour](https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb.py#L298). When a `ScalarFunction` is created it evaluates `fun` and `grad` at an initial point. If that initial point is outside the bounds then a finite difference estimation of grad raises an Exception.

Reply:
I see it's just moved from below